Sources/Apexy/Clients/ConcurrencyClient.swift

```diff
@@ -0,0 +1,57 @@
+//
+//  ConcurrencyClient.swift
+//  
+//
+//  Created by Aleksei Tiurnin on 16.08.2022.
+//
+
+import Foundation
+
+@available(macOS 10.15, iOS 13, watchOS 6, tvOS 13, *)
+public protocol ConcurrencyClient: AnyObject {
+    /// Send request to specified endpoint.
+    /// - Parameters:
+    ///    - endpoint: endpoint of remote content.
+    /// - Returns: response data from the server for the request.
+    func request<T>(_ endpoint: T) async throws -> T.Content where T: Endpoint
+    
+    /// Upload data to specified endpoint.
+    /// - Parameters:
+    ///    - endpoint: endpoint of remote content.
+    /// - Returns: response data from the server for the upload.
+    func upload<T>(_ endpoint: T) async throws -> T.Content where T: UploadEndpoint
+    
+}
+
+@available(macOS 10.15, iOS 13, watchOS 6, tvOS 13, *)
+public extension Client where Self: ConcurrencyClient {
+    
+    func request<T>(_ endpoint: T) async throws -> T.Content where T: Endpoint {
+        try await AsyncAwaitHelper.adaptToAsync(dataTaskClosure: { continuation in
+            request(endpoint, completionHandler: continuation.resume)
+        })
+    }
+    
+    func upload<T>(_ endpoint: T) async throws -> T.Content where T: UploadEndpoint {
+        try await AsyncAwaitHelper.adaptToAsync(dataTaskClosure: { continuation in
+            upload(endpoint, completionHandler: continuation.resume)
+        })
+    }
+}
+
+@available(macOS 10.15, iOS 13, watchOS 6, tvOS 13, *)
+public enum AsyncAwaitHelper {
+    public typealias ContentContinuation<T> = CheckedContinuation<T, Error>
+
+    public static func adaptToAsync<T>(dataTaskClosure: (ContentContinuation<T>) -> Progress) async throws -> T {
+        let progressWrapper = ProgressWrapper()
+        return try await withTaskCancellationHandler(handler: {
```

Comment:
If the request was canceled before the start, `handler` will be called first and then `operation`. In this case, the request will still be launched, even if the cancellation was made first

A small example with cancellation

```
@main
struct App {
    static func main() async {
        do {
            let task = Task<String, Error>(priority: .background) {
                try await fetch()
            }
            task.cancel()
            let value = try await task.value
            print(value)
        } catch {
            print(error)
        }
    }

    static func fetch() async throws -> String {
        await withTaskCancellationHandler {
            print("start")
            return "success"
        } onCancel: {
            print("cancel")
        }

    }
}
```